rust/worker/src/segment/distributed_hnsw_segment.rs

```diff
@@ -172,57 +173,57 @@ impl DistributedHNSWSegmentWriter {
     }
 }
 
-impl SegmentWriter for DistributedHNSWSegmentWriter {
-    fn apply_materialized_log_chunk(
+impl<'a> SegmentWriter<'a> for DistributedHNSWSegmentWriter {
+    async fn apply_materialized_log_chunk(
         &self,
-        records: crate::execution::data::data_chunk::Chunk<super::MaterializedLogRecord>,
-    ) {
-        for record in records.iter() {
-            match record.0.log_record.record.operation {
+        records: crate::execution::data::data_chunk::Chunk<super::MaterializedLogRecord<'a>>,
+    ) -> Result<(), ApplyMaterializedLogError> {
+        for (record, _) in records.iter() {
+            match record.final_operation {
+                // If embedding is not found in case of adds it means that user
+                // did not supply them and thus we should return an error as
+                // opposed to panic.
                 Operation::Add => {
-                    let segment_offset_id = record.0.segment_offset_id;
-                    let embedding = record.0.log_record.record.embedding.as_ref().unwrap();
-                    self.index
-                        .read()
-                        .add(segment_offset_id as usize, &embedding);
+                    let embedding = match record.final_embedding {
+                        Some(e) => e,
+                        None => match record.data_record.as_ref() {
+                            Some(record) => record.embedding,
+                            None => {
+                                tracing::error!("Embedding not set for record {:?}", record);
+                                return Err(ApplyMaterializedLogError::EmbeddingNotSet);
```

Comment:
this will fail the compaction right? how do we want to think of poisoned messages like this?

Reply:
Trim the log from the end, restart compaction.